src/installer/destinations.py

```diff
@@ -55,3 +61,59 @@ def finalize_installation(self, scheme, record_file_path, records):
 
         """
         raise NotImplementedError
+
+
+class SchemeDictionaryDestination(WheelDestination):
+    """file-system destination based on a scheme dictionary."""
+
+    def __init__(
+        self,
+        scheme_dict,
+        interpreter,
+        script_kind,
+        optimise=(1,),
+        hash_algorithm="sha256",
+    ):
+        # type: (Dict[str, str], str, LauncherKind, Tuple[int, ...], str) -> None
+        """Construct destination."""
+        self.scheme_dict = scheme_dict
+        self.interpreter = interpreter
+        self.script_kind = script_kind
+        self.optimise = optimise
+        self.hash_algorithm = hash_algorithm
+
+    def _write_file(self, scheme, path, stream):
+        target_path = os.path.join(self.scheme_dict[scheme], path)
+        # open(..., "x") is not supported in Python 2 so let's check if a file is there ourselves
+        if os.path.exists(target_path):
+            raise FileExistsError(
+                "Target file already exists in the file-system: {}".format(target_path)
+            )
+        with open(target_path, "wb") as f:
+            hash_, size = copyfileobj_with_hashing(stream, f, self.hash_algorithm)
+        return RecordEntry(path, Hash(self.hash_algorithm, hash_), size)
+
+    def write_file(self, scheme, path, stream):
+        # type: (Scheme, FSPath, BinaryIO) -> RecordEntry
+        """Write a file to file-system."""
+        if scheme == "scripts":
+            stream = fix_shebang(stream, self.interpreter)
+        return self._write_file(scheme, path, stream)
+
+    def write_script(self, name, module, attr, section):
+        # type: (str, str, str, ScriptSection) -> RecordEntry
+        """Write an entrypoint script to the file-system."""
+        script = Script(name, module, attr, section)
+        name, data = script.generate(self.interpreter, self.script_kind)
+        return self._write_file("scripts", name, io.BytesIO(data))
+
+    def finalize_installation(self, scheme, record_file_path, records):
+        # type: (Scheme, FSPath, Iterable[RecordEntry]) -> None
+        """Write the RECORD file and generate the Python cache."""
+        record = RecordEntry("RECORD", None, None)
+        self._write_file(
+            scheme,
+            record_file_path,
+            construct_record_file(list(records) + [record]),
```

Comment:
This BytesIO needs to be closed.

I'd make the helper into a context manager with contextlib. The

Reply:
The other option is to clearly state that the end user is responsible for closing this stream, which is also fine by me.

Reply:
Why does it need to be closed by us? I would assume otherwise.

```python
# source
with open('some_file') as f:
    destinatination.write_file('data', 'some_file', f)
```

Reply:
`WheelSource.get_contents` returns an iterable, I would assume it would be implemented like this:

```python
def get_contents(self):
    for file in ...:
        with open(file) as stream:
            yield ..., stream
```

Reply:
`construct_record_file`'s return value is not being closed anywhere.

Reply:
I fixed that.